Lib/test/regrtest.py

```diff
@@ -1910,6 +1938,70 @@ def getexpected(self):
         assert self.isvalid()
         return self.expected
 
+
+if threading is not None:
+    class WatchDogThread(threading.Thread):
+        daemon = True
+
+        def __init__(self, timeout):
+            threading.Thread.__init__(self)
+            self.timeout = timeout
+            self.stream = sys.__stdout__
+            self.quit = False
+            self.reset()
+
+        def reset(self):
+            self.deadline = time.time() + self.timeout
+
+        def write(self, line):
+            self.stream.write(line + "\n")
+            self.stream.flush()
+
+        def run(self):
+            while True:
+                # the sleep duration impacts the delay of .join()
+                # called by .stop()
+                time.sleep(0.1)
+                if self.quit:
+                    return
+                if time.time() >= self.deadline:
+                    break
+
+            try:
+                self.dump_threads()
+            except:
+                self.write("FAILED TO DUMP THREADS")
+            os._exit(1)
+
+        def stop(self):
+            if not self.is_alive():
+                return
+
+            print("Stop watchdog")
+            self.quit = True
+            self.join()
+
+        def dump_thread(self, stack):
+            for filename, lineno, name, line in traceback.extract_stack(stack):
+                self.write('File: "%s", line %d, in %s' % (filename, lineno, name))
+                line = line.strip()
+                if line:
+                    self.write("  %s" % line)
+
+        def dump_threads(self):
+            self.write("*** STACKTRACE - START ***")
```

Comment:
For the case if the previous outputted line is not ended with a newline and for attracting an attention add a `\n` or two before the message.